modules/telemetry/telemetry/include/hephaestus/telemetry/struclog.h

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <iomanip>
+#include <source_location>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <utility>
+#include <vector>
+
+namespace heph::telemetry {
+
+enum class Level : std::uint8_t { TRACE, DEBUG, INFO, WARN, ERROR, FATAL };
+auto operator<<(std::ostream& /*os*/, const Level& /*level*/) -> std::ostream&;
+
+template <typename T>
+concept NotQuotable = !std::convertible_to<T, std::string> && requires(std::ostream& os, T t) {
+  { os << t } -> std::same_as<std::ostream&>;
+};
+
+template <typename T>
+concept Quotable = std::convertible_to<T, std::string>;
+
+template <typename T>
+struct Field final {
+  std::string key;
+  T val;
+};
+
+/**
+ * @brief A class that allows easy composition of logs for structured logging.
+ *        Example(see also struclog.cpp):
+ *        namespace ht=heph::telemetry;
+ *        log(ht::LogEntry{Level::INFO, "my module", "adding"} | "id"_f(12345) | ht::Field{"tag", "test"}) ;
+ *
+ *        logs
+ *        'level=info hostname=goofy location=struclog.h:123 thread-id=5124 time=2023-12-3T8:52:02+0
+ * module="my module" message="adding" id=12345 tag="test"'
+ *
+ *        Remark: We would like to use libfmt with quoted formatting as proposed in
+ *                https://github.com/fmtlib/fmt/issues/825 once its included (instead of sstream).
+ *
+ */
+struct LogEntry {
+  using Fields = std::vector<Field<std::string>>;
+  using Clock = std::chrono::system_clock;
+
+  LogEntry(Level level, std::string&& module, std::string&& message,
+           std::source_location location = std::source_location::current());
+
+  /**
+   * @brief General loginfo consumer, should be used like LogEntry("my message") | Field{"field", 1234}
+   *        Converted to string with stringstream.
+   *
+   * @tparam T any type that is consumeable by stringstream
+   * @param key identifier for the logging data
+   * @param val value of the logging data
+   * @return LogEntry&&
+   */
+  template <NotQuotable T>
+  auto operator|(const Field<T>& field) -> LogEntry&& {
```

Comment:
Would it make sense to use the << operator instead of the | operator? Functionality wise it seems to match the << operator of streams.

Reply:
+1 for operator<<

Reply:
Yes we discussed this and also I proposed it in the original PR #126

For me there is a few arguments for the piping (from strong to weak)
A. in the context of logging << is usually used with streams like ´´´LOG(INFO) << ´´´ or ´´´std::cout<< ´´´ and it feels strange to me to use this in a more functional logging style like ´´´log( ... << ...) ´´´ this might be confusing for people that are used to the first so its better to use sth that are not used to in the context.
B. since ranges | is also used in c++ for chaining elements, although I agree that the meaning is different since its usually modifiers on existing data
C. to me it gives better visual seperation which is quite important for readability

that being said the arguments all are medium strong so I am also open to change to <<
WDYT?

Reply:
> A. in the context of logging << is usually used with streams like ´´´LOG(INFO) << ´´´ or ´´´std::cout<< ´´´ and it feels strange to me to use this in a more functional logging style like ´´´log( ... << ...) ´´´ this might be confusing for people that are used to the first so its better to use sth that are not used to in the context.

One example where logging is done with << inside function calls is https://github.com/catchorg/Catch2/blob/devel/docs/logging.md, so it's totally uncomming.

In the end it's something subjective. I still prefer <<, but I also think that your opinion should be given more weight, because you're the author of it, so if you prefer |, then I'm fine with that too.

Reply:
based on this https://github.com/olympus-robotics/hephaestus/pull/201#discussion_r1837955221
I think we are going to drop both `|` or `<<` and go for variadic